# Variant discovery using samtools. 

## Introduction

This section of the tutorial introduced variant calling using the methods implemented in `samtools`. We will assume that you have completed Part 1 and have available QC'ed and processed sequence alignment files in bam format. 

Steps here will use the following software packages:

- [samtools](http://www.htslib.org/doc/samtools.html)

Each major step has an associated bash script tailored to the UConn CBC Xanadu cluster with appropriate headers for the [Slurm](https://slurm.schedmd.com/documentation.html) job scheduler. The code can easily be modified to run interactively, or in other contexts. 

## Contents
  
1.    [ Update your working directory ](#Update-your-working-directory)  
2.    [ Generate a pileup file with samtools mpileup ](#Generate-a-pileup-file-with-samtools-mpileup)  
3.    [ Call variants with bcftools ](#Call-variants-with-bcftools)  
4.    [ Filter variants with bcftools ](#Filter-variants-with-bcftools)  


## Update your working directory

## Generate a pileup file with mpileup

## Call variants with bcftools

## Filter variants with bcftools